parlai/core/torch_generator_agent.py

```diff
@@ -1655,15 +1799,34 @@ def __init__(self, k, *args, **kwargs):
         super().__init__(*args, **kwargs)
         self.k = k
 
-    def select_paths(self, logprobs, prior_scores, current_length):
+    def select_paths(self, logprobs, prior_scores, current_length) -> _PathSelection:
         values, indices = logprobs.topk(self.k, dim=-1)
         probs = torch.softmax(values, dim=-1)
         choices = torch.multinomial(probs, 1)[:, 0]
         hyp_ids = torch.arange(logprobs.size(0)).to(logprobs.device)
         tok_ids = indices[hyp_ids, choices]
         scores = values[hyp_ids, choices]
         best_scores = prior_scores.expand_as(scores) + scores
-        return (hyp_ids, tok_ids, best_scores)
+
+        vocab_size = logprobs.size()[1]
+        ranks = logprobs.scatter(
+            1,
+            logprobs.argsort(descending=True),
+            torch.arange(vocab_size)
+            .type(torch.FloatTensor)
+            .to(logprobs.device)[None, :]
+            .repeat((self.beam_size, 1))
+            + 1,
+        )
+        tok_ranks = torch.gather(ranks, 1, tok_ids[None, :]).view(-1)
```

Comment:
also here you don't actually need to go through the trouble of resorting the logprobs. `torch.topk` actually returns the values and indices of teh top-k log probs in sorted order for you already, so you can make use of that